xaitk_saliency/interfaces/perturb_image.py

```diff
@@ -0,0 +1,53 @@
+import abc
+from typing import List, Tuple
+
+import numpy as np
+import PIL.Image
+from smqtk_core import Plugfigurable
+
+
+class PerturbImage (Plugfigurable):
+    """
+    Interface abstracting the behavior of taking a reference image and
+    generating some number perturbations of the image along with paired mask
+    matrices indicating where perturbations have occurred and to what amount.
+
+    Implementations should impart no side effects upon the input image.
+    """
+
+    @abc.abstractmethod
+    def perturb(
+        self,
+        ref_image: PIL.Image.Image
+    ) -> Tuple[List[PIL.Image.Image], np.ndarray]:
+        """
+        Transform an input reference image into a number of perturbed
+        variations along with mask matrices indicating the perturbed regions.
+
+        Output images should have the same shape as the input reference image,
+        including channels.
+
+        Output mask matrices should be congruent in length to the number of
+        perturbed images output, as well as share the same height and width
+        dimensions.
+        These masks should indicate the regions in the corresponding perturbed
+        image that has been modified.
+        Values should be in the [0, 1] range, where a value closer to 1.0
+        indicate areas of the image that are *unperturbed*.
+        Mask values should be of a floating-point type in order to allow for
+        fractional perturbation.
```

Comment:
Just a quick thought- when using masks, having them be binary can be pretty useful (i.e. less memory and less storage if cached). Do we plan to support binary masks or should they always be float?

Reply:
That's a good point. Maybe a better way to phrase this is that masks "may be float-valued in order allow for fractional perturbation." The important bit is the value range of [0,1] I think. As long as that promise is made/adhered-to, then maybe the value representation is not important. We could show in examples that where appropriate, numpy/tensor type casting should be performed (which is generally a no-op if the input is already the type requested).

Reply:
Sounds good to me!